### DISCUSSION TITLE
How to set cacheTime for setQuery methods

### DISCUSSION THREAD

**Discussion by push1kb** [TanStack/query#3766] [Q&A]
I want set cacheTime to a query set with `setQuery` method.

```ts
queryClient.setQueryData(key, value, { cacheTime: Infinity });
//                                           ^ throws a type error
```

Upon looking into the `types.d.ts` file it only these

```ts
export interface SetDataOptions {
    updatedAt?: number;
}
```

How to do I add cacheTime and other options to `setQuery`, is there any reason why this was omitted?

**Comment by TkDodo** [comment]
What problem are you trying to solve - what's your use-case?

**Reply by push1kb** [reply]
Since I don't have any redux, and I want to store few strings for all components to access. Hence I'm storing it in React Query. But after 5 minutes, it gets garbage collected and not able to access it

**Reply by TkDodo** [reply]
it gets garbage collected after 5 minutes if you are not using it. You can increase the cacheTime globally, per query-key, or on the `useQuery` instances. I would not recommend to store arbitrary, synchronous data in react-query. It is an async state manager. For synchronous state, use something else like [zustand](https://github.com/pmndrs/zustand).

**Reply by push1kb** [reply]
Reason for not storing in Zustand is that, I have to dehydrate this in Server Side and rehydrate in the Client side. Since I'm perfetching APIs in SSR, passing these values in RQ was the perfect option. 

Setting `cacheTime: Infinity` globally is a workaround for now. But I want to set cacheTime to infinity only for these variables, so it doesn't get lost and my RQ doesn't store old unused data in the cache

**Reply by TkDodo** [reply]
you can set them per query key:
`queryClient.setQueryDefaults(['todos'], { cacheTime: Infinity })`

https://react-query.tanstack.com/reference/QueryClient#queryclientsetquerydefaults

**Reply by push1kb** [reply]
Unfortunately, query defaults are not included in dehydrated state. Guess there's no point in asking RQ to do what it's not meant to do.

I will get the dehydrated state and store in Zustand or Local Storage in Client Side. 

Thanks for your time man.

**Reply by TkDodo** [reply]
only data is part of the dehydrated state, meta information like cacheTime is never included.
but there's a new, different queryClient on the frontend where you hydrate _into_. That's where you have to set the defaults.

**Reply by push1kb** [reply]
Wouldn't it be cool if there's a flag to merged query default options when hydrating? So that we get options from old query client also?

**Reply by TkDodo** [reply]
query options contain functions as well, and setting cacheTime on the server has a different effect than on the client. I don't think we should merge those.
